src/include/sof/audio/crossover/crossover.h

```diff
@@ -142,14 +143,101 @@ static inline crossover_split crossover_find_split_func(int32_t num_sinks)
 	return crossover_split_fnmap[num_sinks - CROSSOVER_2WAY_NUM_SINKS];
 }
 
+static int32_t iir_df2t_2stage_biquads(struct iir_state_df2t *iir, int32_t x)
+{
+	int32_t in;
+	int32_t tmp;
+	int32_t tmp_out;
+	int64_t acc;
+	int32_t out = 0;
+	int d = 0; /* Index to delays */
+	int c = 0; /* Index to coefficient a2 */
+
+	/* Bypass is set with number of biquads set to zero. */
+	if (!iir->biquads)
+		return x;
+
+	/* Coefficients order in coef[] is {a2, a1, b2, b1, b0, shift, gain} */
+
+	/* First-stage biquad filter */
+
+	/* Compute output: Delay is Q4.60
+	 * Coef(Q2.30) x Input(Q1.31) = Q3.61 shift to Q4.60 for accumulating
+	 * Then saturate to Q4.28
+	 */
+	in = x;
+	acc = iir->delay[d];
+	acc += Q_MULTSR_32X32((int64_t)iir->coef[c + 4], in, 30, 31, 60); /* b0 */
+	tmp = sat_int32((int32_t)Q_SHIFT_RND(acc, 60, 28));
+
+	/* Compute first delay */
+	acc = iir->delay[d + 1];
+	acc += Q_MULTSR_32X32((int64_t)iir->coef[c + 3], in, 30, 31, 60); /* b1 */
+	acc += Q_SHIFT_LEFT(((int64_t)iir->coef[c + 1]) * tmp, 58, 60); /* a1 */
+	iir->delay[d] = acc;
+
+	/* Compute second delay */
+	acc = Q_MULTSR_32X32((int64_t)iir->coef[c + 2], in, 30, 31, 60); /* b2 */
+	acc += Q_SHIFT_LEFT(((int64_t)iir->coef[c]) * tmp, 58, 60); /* a2 */
+	iir->delay[d + 1] = acc;
+
+	/* Apply gain Q2.14 x Q4.28 -> Q6.42 */
+	acc = ((int64_t)iir->coef[c + 6]) * tmp; /* Gain */
+
+	/* Apply biquad output shift right parameter
+	 * simultaneously with Q6.42 to Q6.26 conversion and suturation.
+	 */
+	acc = Q_SHIFT_RND(acc, 42 + iir->coef[c + 5], 26);
```

Comment:
The problem with Q6.26 output from first biquad is that it's not feasible for generic (IIR EQ) biquad. This will need a separate optimized HiFi3 version. You would achieve the same down-scaling and headroom for above >1.0 output by using scaling of biquads and apply the gain at last biquad.

Reply:
This function is no more applicable.